### PR TITLE
Support const entries in P4TC

### PR DESCRIPTION
**Pull request by komaljai**
Ticket - https://github.com/p4tc-dev/p4tc-e2e/issues/30

As an example, in the following program:

```
table calculate {
        key = {
            hdr.p4calc.op        : exact @name("op");
        }
        actions = {
            operation_add;
            operation_sub;
            operation_and;
            operation_or;
            operation_xor;
            operation_drop;
        }
        const default_action = operation_drop();
        const entries = {
            P4CALC_PLUS : operation_add();
            P4CALC_MINUS: operation_sub();
            P4CALC_AND  : operation_and();
            P4CALC_OR   : operation_or();
            P4CALC_CARET: operation_xor();
        }
    }
```

The compiler should generate the template code for these constant table enties.
For the above example, it should be the following:

```
$TC p4template update table/calc/MainControlImpl/calculate entry op 0x2b action calc/MainControlImpl/operation_add
$TC p4template update table/calc/MainControlImpl/calculate entry op 0x2d action calc/MainControlImpl/operation_sub
$TC p4template update table/calc/MainControlImpl/calculate entry op 0x26 action calc/MainControlImpl/operation_and
$TC p4template update table/calc/MainControlImpl/calculate entry op 0x7c action calc/MainControlImpl/operation_or
$TC p4template update table/calc/MainControlImpl/calculate entry op 0x5e action calc/MainControlImpl/operation_xor
```

### CODE REVIEW

File: backends/tc/backend.cpp [hunk 1]
```diff
@@ -271,6 +271,53 @@ void ConvertToBackendIR::postorder(const IR::P4Action *action) {
...
+                do {
+                    const int digit = static_cast<int>(static_cast<big_int>(kValue % kBase));
+                    kValue = kValue / kBase;
+                    buf.push_front(Util::DigitToChar(digit));
+                } while (kValue > 0);
+                for (auto ch : buf) value << ch;
+                key = value.str().c_str();
+            }
+            keyList.emplace(keyString, key);
+        }
```
Comment (psivanup) [inline backends/tc/backend.cpp:327]: no other expressions supported?
range, mask etc

File: backends/tc/ebpfCodeGen.cpp [hunk 2]
```diff
@@ -178,6 +178,7 @@ void PNAArchTC::emitParser(EBPF::CodeBuilder *builder) const {
     builder->newline();
     builder->newline();
     emitInstances(builder);
+    // EBPF::EBPFHashAlgorithmTypeFactoryPSA::instance()->emitGlobals(builder);
```
Comment (psivanup) [inline backends/tc/ebpfCodeGen.cpp]: shall remove this line

File: backends/tc/tc.def [hunk 3]
```diff
@@ -288,6 +292,19 @@ class TCTable {
             }
             tcTable += " action " + defaultMissAction->getName();
         }
+        if (const_entries.size() != 0) {
+            for (auto entry : const_entries) {
+                tcTable += "\n$TC p4template create table/" + pipelineName
+                      + "/" + controlName + "/" + tableName
+                      + " entry";
+                for(auto k : entry.second) {
+                    tcTable += " " + k.first + " " + k.second;
```
Comment (psivanup) [inline backends/tc/tc.def]: Is k.second hex digit string? if so, shouldn't it be prefixed with 0x?

File: testdata/p4tc_samples_outputs/calculator.template [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	type exact \
+	keysz 8 nummasks 8 tentries 2048 \
+	table_acts act name calculator/MainControlImpl/operation_add \
+	act name calculator/MainControlImpl/operation_sub \
+	act name calculator/MainControlImpl/operation_and \
+	act name calculator/MainControlImpl/operation_or \
+	act name calculator/MainControlImpl/operation_xor \
+	act name calculator/MainControlImpl/operation_drop
+$TC p4template update table/calculator/MainControlImpl/calculate default_miss_action permissions 0x1024 action calculator/MainControlImpl/operation_drop
+$TC p4template create table/calculator/MainControlImpl/calculate entry op 2b action calculator/MainControlImpl/operation_add permissions 0x1024
```
Comment (psivanup) [inline testdata/p4tc_samples_outputs/calculator.template:37]: Is 2b hex digit string? if so, shouldn't it be prefixed with 0x?

**Review by psivanup — COMMENTED** [review]
please check

File: backends/tc/tc.def (hunk 3 above)
Comment (komaljai) [reply backends/tc/tc.def]: no not required. The type will be extracted from introspection json file

File: testdata/p4tc_samples_outputs/calculator.template (hunk 4 above)
Comment (komaljai) [reply testdata/p4tc_samples_outputs/calculator.template:37]: no type will be extracted from json file.

File: backends/tc/backend.cpp (hunk 1 above)
Comment (komaljai) [reply backends/tc/backend.cpp:327]: addressed

File: backends/tc/ebpfCodeGen.cpp (hunk 2 above)
Comment (komaljai) [reply backends/tc/ebpfCodeGen.cpp]: addressed

File: backends/tc/backend.cpp (hunk 1 above)
Comment (psivanup) [reply backends/tc/backend.cpp:327]: Hope you checked all possible values

**Review by psivanup — APPROVED** [review]
Looks fine

**Comment by komaljai** [comment]
Merged to main

**Pull request closed without merging**